Implement period-tracking component

frenzibyte opened the pull request:
To be used in #8632 

`PeriodTracker` is a component used to allow tracking whether a time value falls within any of the added periods to the tracker, this has been written to allow using that logic in any component that need it, currently used in `BreakTracker` and also to-be-used for #8632.

## Notes
- The `BreakTracker.CurrentBreakIndex` has been removed due to no longer being useful, although it turns out to be used in one of the test cases in `TestSceneBreakTracker`, but it didn't really have a purpose there so I've removed it, if anything we could potentially add an `out` parameter in the checking method providing the index of the interval if a value is inside one.
- The `BreakTracker.Breaks` getter has been removed due to it causing redundancy to the logic after using `IntervalList<double>`, so I've removed it and changed its usages to take the breaks directly from the beatmap instead. Should likely remove the getter from any component that provides one for no reason at some point.

Commit:
Implement interval list

Commit:
Add tests covering most cases of interval list checking

Commit:
Use IntervalList<double> for tracking break periods

bdach commented:
why is this a full collection with add/remove instead of a simple readonly structure...?

bdach commented on osu.Game/Lists/IntervalList.cs:
```diff
@@ -0,0 +1,128 @@
+    /// Represents a list of intervals that can be used for whether a specific value falls into one of them.
+    /// </summary>
+    /// <typeparam name="T">The type of interval values.</typeparam>
+    public class IntervalList<T> : ICollection<Interval<T>>, IReadOnlyList<Interval<T>>
```
this doesn't make sense for things that are not numbers (there is no such thing as an interval of strings). if C# had traits I'd use that, but it doesn't so why not just use `float` or something?

bdach commented on osu.Game/Lists/IntervalList.cs:
```diff
@@ -0,0 +1,128 @@
+        /// <summary>
+        /// The index of the nearest interval from last <see cref="IsInAnyInterval"/> call.
+        /// </summary>
+        protected int NearestIntervalIndex;
```
why `protected`?

iiSaLMaN replied:
I don't think we should expose such an internal field as `NearestIntervalIndex`, it's used for the search inside the method only, changed to `protected` for testing what the index becomes to ensure that it doesn't break.

bdach replied:
the answer is not to test it. you're testing an implementation detail. test the external behaviour only instead (as in, treat the list as a black box that only has `IsInAnyInterval`).

exposing is wrong design-wise as it breaks encapsulation. this is internal state that should not be visible to inheritors.

bdach commented on osu.Game/Lists/IntervalList.cs:
```diff
@@ -0,0 +1,128 @@
+        /// <param name="interval">The interval to add.</param>
+        public void Add(Interval<T> interval) => intervals.Add(interval);
+
+        /// <summary>
+        /// Removes an existing interval from the list.
+        /// </summary>
+        /// <param name="interval">The interval to remove.</param>
+        /// <returns>Whether the provided interval exists in the list and has been removed.</returns>
+        public bool Remove(Interval<T> interval) => intervals.Remove(interval);
+
+        /// <summary>
+        /// Removes all intervals from the list.
+        /// </summary>
+        public void Clear() => intervals.Clear();
+
+        public void CopyTo(Interval<T>[] array, int arrayIndex) => intervals.CopyTo(array, arrayIndex);
+
+        public bool Contains(Interval<T> item) => intervals.Contains(item);
+
+        public IEnumerator<Interval<T>> GetEnumerator() => intervals.GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        #endregion
+
+        #region IReadOnlyList<Interval<T>>
+
+        public Interval<T> this[int index]
+        {
+            get => intervals[index];
+            set => intervals[index] = value;
+        }
+
+        #endregion
```
why? who's going to ever use these?

iiSaLMaN replied:
implementation of `ICollection<Interval<T>>` and `IReadOnlyList<Interval<T>>`, I guess what your review suggests is to make it so simple that the implementation becomes enough for its usages.

bdach commented on osu.Game/Lists/IntervalList.cs:
```diff
@@ -0,0 +1,128 @@
+
+        public Interval(T start, T end)
+        {
+            bool startLessThanEnd = Comparer<T>.Default.Compare(start, end) < 0;
```
no, don't silently fix things up for users. throw an exception instead. this makes it possible to introduce subtle bugs.

bdach reviewed (changes requested):
way too overdone IMO

iiSaLMaN commented:
> why is this a full collection with add/remove instead of a simple readonly structure...?

I've reduced it enough to satisfy all conversations, but I'm gonna keep the add/remove methods just to actually simplify the process of adding intervals.

Commit:
Revert nearest index accessibility back to private

Implementation details should not be tested.

Commit:
Throw if interval with reversed values added

May reduce in subtle bugs.

Commit:
Limit generic type to IConvertible structures (common number types)

Commit:
Simplify IntervalList implementation enough to work for its usages

bdach commented on osu.Game/Lists/IntervalList.cs:
```diff
@@ -0,0 +1,111 @@
+        private readonly SortedList<Interval<T>> intervals = new SortedList<Interval<T>>((x, y) => type_comparer.Compare(x.Start, y.Start));
+        private int nearestIndex;
+
+        public Interval<T> this[int i]
+        {
+            get => intervals[i];
+            set => intervals[i] = value;
+        }
```
Not required and potentially buggy (consider what happens if you do
```csharp
intervalList[1] = new Interval<double>(15, 20);
```
after construction - will it break? will it be fine? I don't know, but it shouldn't be possible in the first place)

bdach commented on osu.Game/Lists/IntervalList.cs:
```diff
@@ -0,0 +1,111 @@
+        /// <param name="interval">The interval to add.</param>
+        public void Add(Interval<T> interval) => intervals.Add(interval);
+
+        /// <summary>
+        /// Removes an existing interval from the list.
+        /// </summary>
+        /// <param name="interval">The interval to remove.</param>
+        /// <returns>Whether the provided interval exists in the list and has been removed.</returns>
+        public bool Remove(Interval<T> interval) => intervals.Remove(interval);
+
+        /// <summary>
+        /// Removes all intervals from the list.
+        /// </summary>
+        public void Clear() => intervals.Clear();
+
+        public IEnumerator<Interval<T>> GetEnumerator() => intervals.GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
```
not required IMO

bdach reviewed (dismissed):
I'm still of the opinion that `IntervalList` should be a non-generic, read-only, *immutable* data structure.

iiSaLMaN commented:
> I'm still of the opinion that `IntervalList` should be a non-generic, read-only, _immutable_ data structure.

I don't mind making it non-generic, but are you fine with creating another normal list to use for adding calculated intervals to store inside `IntervalList` (which I will rename to `IntervalTracker` fwiw)

bdach commented:
I'm not sure what you're asking. I think the list/tracker should take in a `IEnumerable<Interval>` or similar.

iiSaLMaN commented:
like how the hp no-draining sections are calculated:
```csharp
private readonly PeriodTracker nonDrainingSections = new PeriodTracker();

public override void ApplyBeatmap(IBeatmap beatmap)
{
    ...

   foreach (var b in beatmap.Breaks)
   {
       var start = beatmap.HitObjects.Select(h => h.EndTime).LastOrDefault(e => e < b.StartTime);
       var end = beatmap.HitObjects.Select(h => h.StartTime).FirstOrDefault(s => s > b.EndTime);

       nonDrainingSections.AddPeriod(start, end);
   }
}
```

with no `AddPeriod`:
```csharp
private readonly PeriodTracker nonDrainingSections = new PeriodTracker();

public override void ApplyBeatmap(IBeatmap beatmap)
{
    ...

    nonDrainingSections.Periods = beatmap.Breaks.Select(b =>
    {
        var start = beatmap.HitObjects.Select(h => h.EndTime).LastOrDefault(e => e < b.StartTime);
        var end = beatmap.HitObjects.Select(h => h.StartTIme).FirstOrDefault(s => s > b.EndTime);
        return new Period(start, end);
    }
}
```

*actually nvm, it appears to be not that bad*

smoogipoo commented:
You'll want to revert your last commit. Also I agree with the notion that this should be immutable.

iiSaLMaN commented:
> You'll want to revert your last commit. Also I agree with the notion that this should be immutable.

I would agree that the design for referenced commits are confusing here but what you're seeing is just a reference to this PR from 10badc09d53d5586d065ed49f7e2cae09cfb4e9c, unless I'm mistaking for what you mean by "last commit"?

also I have changes ready for this, just discussing one last thing about the component in the above PR. (where it should reside)

smoogipoo commented:
Ah I misread, my mistake.

bdach commented:
What's the status on this PR? You've mentioned that you had the changes ready but there's been no updates for two weeks.

iiSaLMaN commented:
> What's the status on this PR? You've mentioned that you had the changes ready but there's been no updates for two weeks.

Been busy with other PRs + the windows to macOS switch (which I accidentally wiped all local changes in there), I'll get it ready by this day or tomorrow, sorry for this massive delay.

Commit:
Replace interval collection with a more-specific immutable component

Covers all small changes into one commit:
 - Remove generics and use `double` type instead.
 - Make the component immutable and not enumerable for simplicity of it's worth.
 - Make the component more-specific (to time period tracking)
 - Apply small adjustments to the component

Commit:
Adjust, simplify and make the test code more flexible to some degree

Commit:
Apply changes to the BreakTracker and more adjustment

Commit:
Rename is-in-any check method to a more legible name

Commit:
Merge remote-tracking branch 'upstream/master' into add-interval-list

iiSaLMaN commented:
@bdach Pushed. couldn't get it in time, but I've taken a good look on it and hopefully it is in its best shape to you.

bdach committed:
Make PeriodTracker actually immutable

bdach commented:
I've made changes here (mostly cosmetic, with one major removal of the very strange setter that was there before). I think this is tolerable now.

smoogipoo commented on osu.Game/Screens/Play/BreakTracker.cs:
```diff
@@ -2,40 +2,37 @@
-                CurrentBreakIndex = 0;
+
+                breaks = new PeriodTracker(value.Where(b => b.HasEffect)
+                                                .Select(b => new Period(b.StartTime, b.EndTime - BreakOverlay.BREAK_FADE_DURATION)));
```
Is this supposed to have changed, by the inclusion of the fade duration?

iiSaLMaN replied:
This hasn't been changed, the old logic uses `BreakPeriod.Contains(time)` which checks by `time >= Start && time <= (End - BREAK_FADE_DURATION)`. With the `PeriodTracker` component, I wrote it as you can see. (add periods that `b.HasEffect` with exclusion of the fade duration at the end)

bdach replied:
I kind of mindlessly glanced over that, good spot. This has the potential to crash for "too short" breaks (because `Period` checks if `start < end`).

smoogipoo replied:
In this case it's safe because of the `HasEffect` check before it.

iiSaLMaN replied:
Yes I've thought about that the time I wrote this, `HasEffect` checks if the duration is longer than `650`, so no problem there. I was just hesitant a bit since this logic will actually fail if that check inside the property changes.

I guess it's fine for this exact case.

bdach replied:
`BREAK_FADE_DURATION` is explicitly tied to `MIN_BREAK_DURATION` so it should be pretty obvious something is going on there if it's ever changed.

smoogipoo committed:
Merge branch 'master' into add-interval-list

smoogipoo reviewed (approved):
I think this is fine for the time being. Keeping in mind that `BreakTracker` will likely be merged into `GameplayBeatmap`, and that's probably where the draining "tracker" should also be situated going forward.
